lib/mirage_key.ml

```diff
@@ -23,50 +23,54 @@ open Astring
 module Arg = struct
   include Key.Arg
 
-  module type S = sig
-    include Mirage_runtime.Arg.S
-    val sexp_of_t : t -> Sexplib.Type.t
-  end
+  let from_run s = "Mirage_runtime.Arg." ^ s
 
-  (* Could be improved with metaocaml and/or some reflection.
-     [description] and [m] should not need to be provided.
-  *)
-  let of_module
-      (type t) runtime_conv m (module M: S with type t=t) =
-    let conv = Mirage_runtime.Arg.of_module (module M) in
-    let serialize ppf x =
-      Fmt.pf ppf "(%s.t_of_sexp (Sexplib.Sexp.of_string %S))"
-        m (Sexplib.Sexp.to_string @@ M.sexp_of_t x)
+  let make d m of_string to_string =
+    let parser s = match of_string s with
+      | Error (`Msg m) -> `Error ("Can't parse ip address: "^s^": "^m)
+      | Ok ip -> `Ok ip
+    and serialize ppf t =
+      Fmt.pf ppf "(Ipaddr.%s.of_string_exn %S)" m (to_string t)
+    and pp ppf t =
+      Fmt.string ppf (to_string t)
     in
-    Functoria_key.Arg.conv ~conv ~serialize ~runtime_conv
+    Functoria_key.Arg.conv
+      ~conv:(parser, pp)
+      ~serialize
+      ~runtime_conv:(from_run d)
 
-  let from_run s = "Mirage_runtime.Arg." ^ s
-  let builtin d mn m = of_module (from_run d) mn m
+  module type S = sig
+    type t
+    val of_string : string -> (t, [ `Msg of string ]) result
+    val to_string : t -> string
```

Comment:
The Cmdliner way of defining these is to have a `pp` function here instead of a `to_string`

Reply:
this signature is used by `make` above, for (a) Cmdliner style parser and pretty printer, but also (b) serialise and deserialise between configuration and runtime phase -- in this second phase we use `to_sting` directly.

I find it easier to reason that `\forall x . of_string (to_string x) = Ok x` than to include a pretty printer in that equation (which btw was the initial implementation in this PR, as written in the first comment). But I'm open for discussion why/whether we should use `pp` and `of_string` instead.

Reply:
pp combinators are just more efficient to combine, but I don't have a strong opinion.

Reply:
And you should probably use `Fmt.to_to_string` and `Fmt.of_to_string` to convert between `pp` and `to_string` functions.